src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java

```diff
@@ -92,6 +99,11 @@ public final class BlazeOptionHandler {
     this.commandAnnotation = commandAnnotation;
     this.optionsParser = optionsParser;
     this.invocationPolicy = invocationPolicy;
+    allOptionsClasses = runtime.getCommandMap().values().stream()
+        .map(BlazeCommand::getClass)
+        .flatMap(cmd -> BlazeCommandUtils.getOptions(cmd, runtime.getBlazeModules(),
+            runtime.getRuleClassProvider()).stream())
+        .collect(toImmutableList());
```

Comment:
Good point, we should add a `.distinct()` here to make the cache key smaller.

I don't think that passing to a `Set` would reduce the overall complexity of this PR though: The problem isn't that option *classes* collide, it's that individual options can collide across classes. For example, both `cquery` and `query` define an `--output` flag, but those flag definition are distinct and differ in certain aspects (e.g. the allowed values). However, crucially, both options accept an argument (that is, are neither boolean nor Void options) and hence parse identically, as checked by `equivalentForParsing`. That's why this PR introduces an additional type of equality that is more lenient than `Object#equals` for `OptionDefinition`s.

Reply:
Thanks that wasn't clear to me. Hmm that is an annoying bit of necessary complexity of this FR.

Reply:
I would definitely add more comments as part of polishing this PR if you approve of the approach.

I agree that this additional complexity is annoying, but if it isn't taken care of by Bazel itself, it is pushed onto end users. I actually think that compared to the level of effort required to manually apply flags to all commands they are valid for, the complexity of `OptionsParserImpl#equivalentForParsing` is comparatively contained.

Reply:
> For example, both cquery and query define an --output flag ... However, crucially, both options

What would happen if they weren't equivalent for parsing?

> However, crucially, both options accept an argument (that is, are neither boolean nor Void options) and hence parse identically

What would be wrong with two different options with the same name that were both booleans?

Reply:
I'm not clear on why we're doing the `equivalentForParsing` check, especially because it seems like a very loose check. For example, if `some_option` accepts different values in different commands (such as two different enums), the check passes correct? What are we trying to guard against exactly?